### ISSUE TITLE
bottomSheet will rebuild after each setState

### ISSUE THREAD

**Issue by SakurasDuck** [flutter/flutter#73033]
![test (1) (1)](https://user-images.githubusercontent.com/40283698/103209470-26a3c880-493e-11eb-97f8-c40fec168733.gif)

<details> 
<summary>Code Sample</summary> 

``` 
import 'dart:math';

import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(
    theme: ThemeData(primaryColor: Colors.white),
    debugShowCheckedModeBanner: false,
    title: 'demo',
    home: Main(),
  ));
}

class Main extends StatefulWidget {
  @override
  _Main createState() => _Main();
}

class _Main extends State<Main> {
  int count = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(title: Text("appbar")),
        body: Column(
          children: [
            RaisedButton(
              child: Text("count $count"),
              onPressed: () {
                setState(() {
                  count = count + 1;
                });
              },
            ),
            Test(
              child: Text('Scaffold.body'),
              color: Colors.red,
            ),
            Expanded(child: Container())
          ],
        ),
        bottomSheet: Test(
          child: Text('Scaffold.bottomSheet'),
          color: Colors.green,
        ));
  }
}

class Test extends StatefulWidget {
  final Color color;

  final Widget child;

  @override
  _Test createState() => _Test();

  const Test({@required this.child, this.color, Key key}) : super(key: key);
}

class _Test extends State<Test> {
  double height = 300;

  @override
  void initState() {
    super.initState();
    setState(() {
      height = Random().nextDouble() * height;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: widget.color ?? Colors.green,
      height: height,
      child: Center(
        child: widget.child,
      ),
    );
  }
}
```
</details>

**flutter doctor -v**

<details>

```

 [√] Flutter (Channel stable, 1.22.4, on Microsoft Windows [Version 10.0.18363.1256], locale zh-CN)
    • Flutter version 1.22.4 at D:\Tools\Android\flutter
    • Framework revision 1aafb3a8b9 (6 weeks ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4
    • Pub download mirror http://pub.lan.wiqun.com
    • Flutter download mirror https://storage.flutter-io.cn


[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at D:\Tools\Android\SDK
    • Platform android-30, build-tools 29.0.3
    • ANDROID_HOME = D:\Tools\Android\SDK
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[!] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[!] IntelliJ IDEA Ultimate Edition (version 2019.1)
    • IntelliJ at C:\Users\Pawn\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\191.7479.19
    X Flutter plugin not installed; this adds Flutter specific functionality.
    • Dart plugin version 191.8593
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[√] VS Code (version 1.51.1)
    • VS Code at C:\Users\Pawn\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.16.0

[√] Connected device (1 available)
    • VOG AL00 (mobile) • MQS0219A17009076 • android-arm64 • Android 10 (API 29)


```

</details>

i think Scaffold.didUpdateWidget() to rebuild it,but I don't know why

```
  @override
  void didUpdateWidget(Scaffold oldWidget) {
    // Update the Floating Action Button Animator, and then schedule the Floating Action Button for repositioning.
    if (widget.floatingActionButtonAnimator != oldWidget.floatingActionButtonAnimator) {
      _floatingActionButtonAnimator = widget.floatingActionButtonAnimator ?? _kDefaultFloatingActionButtonAnimator;
    }
    if (widget.floatingActionButtonLocation != oldWidget.floatingActionButtonLocation) {
      _moveFloatingActionButton(widget.floatingActionButtonLocation ?? _kDefaultFloatingActionButtonLocation);
    }
    if (widget.bottomSheet != oldWidget.bottomSheet) {  ///???
      assert(() {
        if (widget.bottomSheet != null && _currentBottomSheet?._isLocalHistoryEntry == true) {
          throw FlutterError.fromParts(<DiagnosticsNode>[
            ErrorSummary(
              'Scaffold.bottomSheet cannot be specified while a bottom sheet displayed '
              'with showBottomSheet() is still visible.'
            ),
            ErrorHint(
              'Use the PersistentBottomSheetController '
              'returned by showBottomSheet() to close the old bottom sheet before creating '
              'a Scaffold with a (non null) bottomSheet.'
            ),
          ]);
        }
        return true;
      }());
      _closeCurrentBottomSheet();
      _maybeBuildPersistentBottomSheet();
    }
    super.didUpdateWidget(oldWidget);
  }
```

@TahaTesser
thanks

**Comment by TahaTesser** [comment]
Hi @SakurasDuck 
This looks like a bug to me, according to https://github.com/flutter/flutter/issues/2115, it shouldn't rebuild
<details> 
<summary>code sample</summary> 

```dart
import 'dart:math';

import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(
    theme: ThemeData(primaryColor: Colors.white),
    debugShowCheckedModeBanner: false,
    title: 'demo',
    home: Main(),
  ));
}

class Main extends StatefulWidget {
  @override
  _Main createState() => _Main();
}

class _Main extends State<Main> {
  int count = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(title: Text("appbar")),
        body: Column(
          children: [
            RaisedButton(
              child: Text("count $count"),
              onPressed: () {
                setState(() {
                  count = count + 1;
                });
              },
            ),
            Test(
              child: Text('Scaffold.body'),
              color: Colors.red,
            ),
            Expanded(child: Container())
          ],
        ),
        bottomSheet: Test(
          child: Text('Scaffold.bottomSheet'),
          color: Colors.green,
        ));
  }
}

class Test extends StatefulWidget {
  final Color color;

  final Widget child;

  @override
  _Test createState() => _Test();

  const Test({@required this.child, this.color, Key key}) : super(key: key);
}

class _Test extends State<Test> {
  double height = 300;

  @override
  void initState() {
    super.initState();
    setState(() {
      height = Random().nextDouble() * height;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: widget.color ?? Colors.green,
      height: height,
      child: Center(
        child: widget.child,
      ),
    );
  }
}

``` 

</details>
<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel dev, 1.26.0-1.0.pre, on macOS 11.1 20C69 darwin-x64, locale en-GB)
    • Flutter version 1.26.0-1.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 63062a6443 (2 weeks ago), 2020-12-13 23:19:13 +0800
    • Engine revision 4797b06652
    • Dart version 2.12.0 (build 2.12.0-141.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[!] Android Studio
    • Android Studio at /Applications/Android/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Android Studio not found at /Applications/Android/Contents
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.0

[✓] Connected device (4 available)
    • RMX2001 (mobile)   • EUYTFEUSQSRGDA6D                     • android-arm64  • Android 10 (API 29)
    • iPhone 12 (mobile) • 12D8FF8E-2815-436E-9951-B8A6A42E4ACF • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)    • macos                                • darwin-x64     • macOS 11.1 20C69 darwin-x64
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 87.0.4280.88

! Doctor found issues in 1 category.
```
```bash
[✓] Flutter (Channel master, 1.26.0-2.0.pre.137, on macOS 11.1 20C69 darwin-x64, locale en-GB)
    • Flutter version 1.26.0-2.0.pre.137 at /Users/tahatesser/Code/flutter_master
    • Framework revision 8f5d0371af (2 days ago), 2020-12-26 11:07:36 +0330
    • Engine revision f5364860ab
    • Dart version 2.12.0 (build 2.12.0-179.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[!] Android Studio
    • Android Studio at /Applications/Android/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Android Studio not found at /Applications/Android/Contents
    • Try updating or re-installing Android Studio.

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.0

[✓] Connected device (4 available)
    • RMX2001 (mobile)   • EUYTFEUSQSRGDA6D                     • android-arm64  • Android 10 (API 29)
    • iPhone 12 (mobile) • 12D8FF8E-2815-436E-9951-B8A6A42E4ACF • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)    • macos                                • darwin-x64     • macOS 11.1 20C69 darwin-x64
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 87.0.4280.88

! Doctor found issues in 1 category.
```

</details>

**Comment by justinmc** [comment]
CC @Piinks

**Comment by xu-baolin** [comment]
Maybe fixed by #73084
